### PR TITLE
Metrics Archives

### PR DESCRIPTION
**Pull request by mvollmer**
Followup pull requests:

- Use metrics channels for other plots
- Scrolling/zooming
- Config UI for the logger

Instructions:

Create `/var/lib/pcp/config/pmlogconf/tools/cockpit` with this content:

```
#pmlogconf-setup 2.0
ident   Metrics used by Cockpit
force   include
        kernel.all.cpu.nice
        kernel.all.cpu.user
        kernel.all.cpu.sys
        mem.util.used
        network.interface.total.bytes
        disk.dev.total_bytes
```

Run this command:

```
/usr/libexec/pcp/bin/pmlogconf /var/lib/pcp/config/pmlogger/config.default
```

and hold down the Return key for about 10 seconds until the shell prompt reappers.  Check that the cockpit metrics have been added.

Start (or restart) pmlogger:

```
systemctl enable pmcd
systemctl enable pmlogger
systemctl start pmcd
systemctl start pmlogger
```

### CODE REVIEW

File: src/bridge/cockpitchannel.c [hunk 1]
```diff
@@ -112,7 +112,9 @@ G_DEFINE_TYPE (CockpitChannel, cockpit_channel, G_TYPE_OBJECT);
 static gboolean
 on_idle_prepare (gpointer data)
 {
-  cockpit_channel_prepare (data);
+  CockpitChannel *self = COCKPIT_CHANNEL (data);
+  cockpit_channel_prepare (self);
+  g_object_unref (self);
```
Comment (stefwalter) [inline src/bridge/cockpitchannel.c]: If we are trying to have a guard across the prepare call we should do that right here and not split it across the idle call. The prepare_tag is already guarded by a g_source_remove().

File: src/bridge/cockpitpcpmetrics.c [hunk 2]
```diff
@@ -47,14 +47,24 @@ typedef struct {
 } MetricInfo;
 
 typedef struct {
+  int context;
+  gint64 start;
+} ArchiveInfo;
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: Why are we accessing multiple archive files in one channel?

**Comment by stefwalter** [comment]
Why do we need to write our own janitor? Why doesn't pmlogger do this? Well it's stupid that PCP isn't using round robin files to start off with.

**Comment by stefwalter** [comment]
Shouldn't we be using pmlogger_daily for log rotation?

**Comment by stefwalter** [comment]
Given the massive amount of machinery that PCP requires in order to function, we should just configure the main pmlogger when we are requested to do so, and look in the standard location for archives.

**Comment by fche** [comment]
stefwalter, "Why do we need to write our own janitor?"

To some extent, that's the end point if one goes down the road of managing one's own pmlogger stable.  But you don't have to go that far.  Remember, we talked about y'all using the system pmlogger (with some customized configuration if necessary - drop bits into /var/lib/pcp/config/pmlogconf/... e.g.), in which case it'll be rotated for you via cron jobs.  We also talked about y'all using pmmgr (with some customized configuration if necessary), in which case pmmgr will rotate things for you (without cron).  I'd be happy to elaborate upon either of those paths.

**Comment by stefwalter** [comment]
@fche agree.

In addition we have all the machinery we need to adapt to people removing or changing the configuration, and just not showing certain graph data if they do. Or pulling from an archive with a different interval etc.

File: src/bridge/cockpitpcpmetrics.c (hunk 2 above)
Comment (mvollmer) [inline src/bridge/cockpitpcpmetrics.c]: Because pmlogger will create a new archive every time it starts and I think it is a nice service of the metrics channel to put them back together again.

Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: Isn't that the job of pmlogger_merge?

Comment (fche) [inline src/bridge/cockpitpcpmetrics.c]: Given that you'll probably have at least two archives (one recent and one currently-written-to), this sort of gluing-together would be helpful & nice.  PMAPI will provide exactly this service within a month or two.  (It doesn't mean that individual archive boundaries must necessarily be communicated to the gui.)

pmmgr can keep all prior (but age-limited) data within a single archive (in addition to the currently-written-to one), but it turns out that repeatedly merging/slicing archives together into one tickles a libpcp bug http://oss.sgi.com/bugzilla/show_bug.cgi?id=1044 so we've switched defaults to 'granular' mode, so that N period-aligned archives are kept)

File: src/bridge/cockpitchannel.c (hunk 1 above)
Comment (mvollmer) [inline src/bridge/cockpitchannel.c]: Done.  I split this commit out into its own pull request.

**Comment by mvollmer** [comment]
> Remember, we talked about y'all using the system pmlogger (with some customized configuration if necessary - drop bits into /var/lib/pcp/config/pmlogconf/... e.g.)

Yes, I actually like this option.  It would be nice to be able to tell people "and you can also change the raw pcp configuration and Cockpit will adapt".

But this pull request here is as large a step as I want to make as the first one.

**Comment by mvollmer** [comment]
Depends on #1699.

**Comment by mvollmer** [comment]
> But this pull request here is as large a step as I want to make as the first one.

Right, changed my mind.  Cockpit-logger is gone and we now use the default archives.

@fche, could you check the instructions above for how to add the Cockpit metrics to the default pmlogger configuration?  How would one do this programmatically?

**Comment by mvollmer** [comment]
Okay, this should be ready for another round of review.

**Comment by mvollmer** [comment]
Didn't do the valgrinding yet...

**Comment by stefwalter** [comment]
@andreasn could you look at the initial UI changes here? Do they fit inline with what you're planing on graphing archived resource data? Usually we end up living with this stuff for a long time even though we know it's temporary.

**Comment by fche** [comment]
@mvollmer, OK, if you really want to use the system "primary" pmlogger, that's indeed roughly the way: deposit a pmlogconf fragment in the right spot; cause regeneration of pmlogger config.default file; restart pmlogger.  (You are free to use a different sampling rate with a "delta NNN" clause in the pmlogconf input.)  You do not need to restart pmcd (as it is unaware of clients such as pmlogger).  Any other service-pmlogger instances that might get configured later will probably also inherit the new pmlogconf bits, for better or for worse, though OTOH that is a manual process (of changing /etc/pcp/pmlogger/control).

**Comment by mvollmer** [comment]
> OK, if you really want to use the system "primary" pmlogger

I need some advise about this.

*  We need to provide a maintenance-free experience for people who just want to use Cockpit and have never heard of PCP and would rather never hear about it.

* We should provide no suprises to people who know PCP but not Cockpit and want to configure it in a detailed way.  They should be able to do that without breaking Cockpit by accident. This includes both system administrators that tweak it directly or via things like Puppet and other software that, like Cockpit, wants to use PCP as a implementation detail.

*  We can make Cockpit reasonably robust so that it can tolerate variations in the configuration (such as the sampling rate, rotation rules, whether or not pmlogger is actually running or not).

* We _might_ add flexibility to Cockpit so that the concrete set of metrics shown in the graphs is actually configurable, with or without UI.  That would be cool, but  further down the road.

@fche, can you think of a perfectly normal PCP configuration change that would break Cockpit if we follow the current approach?  For example, can people move the location of the archives around via tweaks to something in /etc?  Do we need to pick that up in Cockpit?

**Comment by mvollmer** [comment]
> cause regeneration of pmlogger config.default file

@fche, what's the recommended way to do this?  The man page of pmlogconf talks about non-interactive use, but it's not obvious to me how to do that.

**Comment by mvollmer** [comment]
> that's indeed roughly the way: 

@fche, is there a detailed description of "the way" somewhere?

(Thanks a lot for all the effort you put into this, btw!  I really appreciate it.)

File: configure.ac [hunk 3]
```diff
@@ -248,7 +248,7 @@ if test "x$GCC" = "xyes"; then
           -Werror=implicit-function-declaration \
           -Werror=pointer-arith -Werror=init-self \
           -Werror=format-security \
-          -Werror=missing-include-dirs -Werror=aggregate-return $CFLAGS"
+          -Werror=missing-include-dirs $CFLAGS"
```
Comment (stefwalter) [inline configure.ac]: Perhaps a separate commit and commit message with reasoning for this change?

File: pkg/shell/plot.js [hunk 4]
```diff
@@ -110,15 +116,14 @@ var shell = shell || { };
  * == true), or stops hovering over it ('val' == false).
  */
 
-shell.plot = function plot(element, x_range) {
+    shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Accidental indent?

File: pkg/shell/plot.js [hunk 5]
```diff
@@ -160,16 +165,33 @@ shell.plot = function plot(element, x_range) {
         walk_timer = null;
     }
 
-    function reset() {
+    function reset(x_range_seconds, x_stop_seconds) {
         stop_walking();
         for (var i = 0; i < data.length; i++)
             data[i].stop();
 
+        // XXX - breaks with DST change during x_range.
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Why are we dealing with local time at all for calculations? Shouldn't we just convert to local time immediately when displaying the data?

File: pkg/shell/plot.js [hunk 6]
```diff
@@ -303,8 +337,50 @@ shell.plot = function plot(element, x_range) {
                     if (instances[i] !== null)
                         cur_samples[i] = [];
                 }
+
                 if (series_data === null)
-                    init();
+                    init(msg);
+
+                var now_date = new Date();
+                t = (now_date.getTime() - now_date.getTimezoneOffset()*60*1000) + (msg.timestamp - msg.now);
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Need spaces around the ```*``` binary operators.

File: pkg/shell/plot.js [hunk 7]
```diff
@@ -303,8 +337,50 @@ shell.plot = function plot(element, x_range) {
...
+                if (archive_channel && t > timestamp + 2*interval)
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Ditto.

File: pkg/shell/plot.js [hunk 8]
```diff
@@ -303,8 +337,50 @@ shell.plot = function plot(element, x_range) {
...
+            factor = 1;
+
+        archive_channel = cockpit.channel({ payload: "metrics1",
+                                            source: "pcp-archive",
+                                            metrics: metrics,
+                                            instances: desc.instances,
+                                            omit_instances: desc.omit_instances,
+                                            interval: interval,
+                                            timestamp: -x_range-x_offset,
+                                            limit: (x_offset > 0)? (x_range / interval) : undefined,
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Space around binary operator.

File: pkg/shell/plot.js [hunk 9]
```diff
@@ -303,8 +337,50 @@ shell.plot = function plot(element, x_range) {
...
+        else
+            factor = 1;
+
+        archive_channel = cockpit.channel({ payload: "metrics1",
+                                            source: "pcp-archive",
+                                            metrics: metrics,
+                                            instances: desc.instances,
+                                            omit_instances: desc.omit_instances,
+                                            interval: interval,
+                                            timestamp: -x_range-x_offset,
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Space around binary operator.

File: pkg/shell/shell.html [hunk 10]
```diff
@@ -112,6 +112,24 @@
...
+            <button class="btn btn-default" data-seconds="1800">30min</button>
+            <button class="btn btn-default" data-seconds="300">5min</button>
+          </div>
+          <div class="btn-group">
+            <button class="btn btn-default" id="dashboard-scroll-left">&lt;</button>
+            <button class="btn btn-default" id="dashboard-scroll-right">&gt;</button>
+          </div>
+        </div>
+        <br>
+        <br>
```
Comment (stefwalter) [inline pkg/shell/shell.html]: Suggest using proper css class to achieve spacing here.

File: src/bridge/cockpitpcpmetrics.c [hunk 11]
```diff
@@ -112,16 +121,23 @@ build_meta (CockpitPcpMetrics *self,
...
   JsonObject *root;
   pmValueSet *vs;
-  gint64 timestamp;
+  struct timeval now_timeval;
+  gint64 timestamp, now;
   char *instance;
   int i, j;
   int rc;
 
+  gettimeofday (&now_timeval, NULL);
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: The metrics are using monotonic time (see CockpitMetrics). Is it intentional that we use wall time here? Just double checking.

File: src/bridge/cockpitpcpmetrics.c [hunk 12]
```diff
@@ -655,8 +741,7 @@ cockpit_pcp_metrics_prepare (CockpitChannel *channel)
           if (instances)
             {
               pmDelProfile (info->desc.indom, 0, NULL);
-              for (int i = 0; instances[i]; i++)
-                {
+              for (int i = 0; instances[i]; i++)                {
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: Indentation?

File: src/bridge/cockpitpcpmetrics.c [hunk 13]
```diff
@@ -561,60 +684,24 @@ convert_metric_description (CockpitPcpMetrics *self,
...
-    {
-      g_warning ("no \"source\" option specified for metrics channel");
-      goto out;
-    }
-  else if (g_str_equal (source, "direct"))
-    {
-      type = PM_CONTEXT_LOCAL;
-      name = NULL;
-    }
-  else if (g_str_equal (source, "pmcd"))
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: We need to file an issue here. pmNewContext() can block for many 10s of seconds if a pmcd is not responsive. This means connecting to pmcd may need to happen in a thread, which has threading implications.

File: src/bridge/cockpitpcpmetrics.c [hunk 14]
```diff
@@ -675,6 +760,255 @@ cockpit_pcp_metrics_prepare (CockpitChannel *channel)
...
+static gboolean
+add_archive (CockpitPcpMetrics *self,
+             const gchar *name)
+{
+  ArchiveInfo *info;
+  pmLogLabel label;
+  int rc;
+
+  info = g_new0 (ArchiveInfo, 1);
+  info->context = pmNewContext (PM_CONTEXT_ARCHIVE, name);
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: @fche Are there consequences to opening many many archive contexts at once?

File: src/bridge/cockpitpcpmetrics.c [hunk 15]
```diff
@@ -675,6 +760,255 @@ cockpit_pcp_metrics_prepare (CockpitChannel *channel)
...
+    {
+      type = PM_CONTEXT_ARCHIVE;
+      name = g_strdup (source);
+    }
+  else if (g_str_has_prefix (source, "pcp-archive"))
+    {
+      gchar hostname[HOST_NAME_MAX + 1];
+      if (gethostname (hostname, HOST_NAME_MAX) < 0)
+        {
+          g_warning ("Error getting hostname: %m");
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: This is not a "protocol-error".

File: src/bridge/cockpitpcpmetrics.c [hunk 16]
```diff
@@ -675,6 +760,255 @@ cockpit_pcp_metrics_prepare (CockpitChannel *channel)
...
+          goto out;
+        }
+      hostname[HOST_NAME_MAX] = '\0';
+      type = PM_CONTEXT_ARCHIVE;
+      name = g_strdup_printf ("/var/log/pcp/pmlogger/%s", hostname);
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: Ha ha ha ha .... sigh. Oh my.

File: src/bridge/cockpitpcpmetrics.c [hunk 17]
```diff
@@ -675,6 +760,255 @@ cockpit_pcp_metrics_prepare (CockpitChannel *channel)
...
+              gchar *path = g_build_filename (name, entry, NULL);
+              path[strlen(path)-strlen(".meta")] = '\0';
+              add_archive (self, path);
+              g_free (path);
+            }
+        }
+      g_dir_close (dir);
+    }
+  else
+    add_archive (self, name);
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: This needs logging of errors that are not ENOTDIR

File: src/bridge/cockpitpcpmetrics.c [hunk 18]
```diff
@@ -675,6 +760,255 @@ cockpit_pcp_metrics_prepare (CockpitChannel *channel)
...
+  rc = pmGetArchiveLabel (&label);
+  if (rc < 0)
+    {
+      g_message ("%s: couldn't read archive label of %s: %s", self->name, name, pmErrStr (rc));
+      pmDestroyContext (info->context);
+      g_free (info);
+      return FALSE;
+    }
+
+  info->start = label.ll_start.tv_sec*1000 + label.ll_start.tv_usec / 1000;
```
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: Spaces around binary operators.

**Comment by stefwalter** [comment]
Further graph UI design is happening here: #1700

**Comment by andreasn** [comment]
> could you look at the initial UI changes here? Do they fit inline with what you're planing on graphing archived resource data?

Yes, they roughly match what's in #1700 
There is a lot of additional little ui details there, but it seems to head in the right direction.

File: src/bridge/cockpitpcpmetrics.c (hunk 13 above)
Comment (fche) [inline src/bridge/cockpitpcpmetrics.c]: See [man pcpintro], $PMCD_CONNECT_TIMEOUT.

File: src/bridge/cockpitpcpmetrics.c (hunk 14 above)
Comment (fche) [inline src/bridge/cockpitpcpmetrics.c]: Should be fine.  Memory will be consumed for all the .meta / .index content, but that's on the order of kilobytes.

File: src/bridge/cockpitpcpmetrics.c (hunk 13 above)
Comment (stefwalter) [inline src/bridge/cockpitpcpmetrics.c]: > See [man pcpintro], $PMCD_CONNECT_TIMEOUT.

This process is async event driven cannot block for more than about max 100 ms anywhere during normal operation, and less than that is recommended.. If we reduce the timeout to less than 100 ms, then the connection to pmcd will likely race and fail. So we would need to do pmcd access out of process and/or in a thread.

So long running tasks like pmNewContext(PM_CONTEXT_HOST) will need

**Comment by fche** [comment]
Hi -


The main difference over using the system "primary" pmlogger vs. a
private pmlogger or pmmgr service is that in the former case you
relinquish control over details like rotation, restarts, allow your
configuration extensions to leak out to other uses, and force use
collection of pcp-default metrics even though cockpit's not interested
in them.

Your own fine-tuned pmmgr-based service (as I outlined in that other
github comment ... wherever it has gone :-) would isolate you from
unrelated sysadmin choices and minimize cockpit's demands on an
originally-pcp-inactive system.

Neither choice expects much from a pcp-ignorant sysadmin though.


> @fche, can you think of a perfectly normal PCP configuration change that would
> break Cockpit if we follow the current approach? For example, can people move
> the location of the archives around via tweaks to something in /etc? Do we need
> to pick that up in Cockpit?

Apprx. all the pcp directories can in principle be modified by the
$PCP_CONF (default /etc/pcp.conf) file, whose contents may be returned
to an app via pmGetConfig().  All I'd suggest is that instead of
hardcoding paths like /var/lib/pcp, use pmGetConfig("PCP_VAR_DIR").


> > [...] cause regeneration of pmlogger config.default file
> 
> @fche, what's the recommended way to do this?

If you were using your own pmmgr service, it would happen for you
automatically every time pmmgr finds the target pmcd.  If you are
using the system "primary" pmlogger, you will need to run pmlogconf
yourself. (If configured that way, pmmgr runs apprx.
"pmlogconf -c -r -h HOSTNAME OUTPUTFILE".)

Or else, remove the /var/lib/pcp/config/pmlogger/config.default file
and restart the pmlogger service.  It should regenerate that file
just once, then reuse it thereafter.

File: src/bridge/cockpitpcpmetrics.c (hunk 13 above)
Comment (fche) [inline src/bridge/cockpitpcpmetrics.c]: "If we reduce the timeout to less than 100 ms, then the connection to pmcd will likely race and fail."

In my experience, even PMCD_CONNECT_TIMEOUT=0 connections to nearby remote hosts regularly go through fine.

**Comment by mvollmer** [comment]
> Or else, remove the /var/lib/pcp/config/pmlogger/config.default file and restart the pmlogger service.

That seems to be a good option.

**Comment by mvollmer** [comment]
> All I'd suggest is that instead of hardcoding paths like /var/lib/pcp, use pmGetConfig("PCP_VAR_DIR").

Will do, thanks!

**Comment by andreasn** [comment]
I've been running this for more than a day now. One thing I've noticed is that there is in total 8 different time steps and I feel that's a bit overwhelming. Can we cut that down a bit? If the selection control in #1700 is implemented, I think we could do with less steps.
What about:
1 week | 1 day | 3 hours | 30 min

1 week could then show 7 days
1 day could show a stop every 3 hours
3 hours could show a stop every 30 min

File: configure.ac (hunk 3 above)
Comment (mvollmer) [inline configure.ac]: Makes sense: #1756

File: pkg/shell/plot.js (hunk 4 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Indeed.  Fixed.

File: pkg/shell/plot.js (hunk 5 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: This is how flot wants to be used, I think.  But I agree, we should do our own tick label formatting.

For example, flot doesn't show the day when you are zoomed in enough, so when you scroll far enough into the past, you can't tell what day you are looking at.

File: src/bridge/cockpitpcpmetrics.c (hunk 11 above)
Comment (mvollmer) [inline src/bridge/cockpitpcpmetrics.c]: Yes, ther "now" field should be of the same kind as the "timestamp" field.

File: src/bridge/cockpitpcpmetrics.c (hunk 12 above)
Comment (mvollmer) [inline src/bridge/cockpitpcpmetrics.c]: Indeed.

File: src/bridge/cockpitpcpmetrics.c (hunk 14 above)
Comment (mvollmer) [inline src/bridge/cockpitpcpmetrics.c]: I'll still put in a limit of 200 contexts, also to put a limit on the time we spend scanning them.

File: src/bridge/cockpitpcpmetrics.c (hunk 15 above)
Comment (mvollmer) [inline src/bridge/cockpitpcpmetrics.c]: Fixed.

File: src/bridge/cockpitpcpmetrics.c (hunk 16 above)
Comment (mvollmer) [inline src/bridge/cockpitpcpmetrics.c]: I'll use pmGetConfig for this.

File: src/bridge/cockpitpcpmetrics.c (hunk 18 above)
Comment (mvollmer) [inline src/bridge/cockpitpcpmetrics.c]: Done.

File: pkg/shell/plot.js (hunk 9 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Done.

File: pkg/shell/plot.js (hunk 8 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Done.

File: pkg/shell/plot.js (hunk 7 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Done.

File: pkg/shell/plot.js (hunk 6 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Done.

**Comment by mvollmer** [comment]
> > All I'd suggest is that instead of hardcoding paths like /var/lib/pcp, use
> > pmGetConfig("PCP_VAR_DIR").
>
> Will do, thanks!

From pmGetConfig docs:

```
If the variable is not found in either the environment or the PCP configuration file (or the PCP configuration file is not found and PCP_CONF is not set in the environment), then a fatal error message is printed and the process will exit. Although this sounds drastic, it is the only course of action available because the PCP configuration or installation is fatally flawed. 
```

Right, can't use that, can we?

**Comment by mvollmer** [comment]
> Right, can't use that, can we?

On the other hand, I'm pretty sure that libpcp uses pmGetConfig internally, so we are victim to this behavior anyway.

Also, only a missing config file seems to cause an abort.  A undefined variable just returns "".

@fche?

File: src/bridge/cockpitpcpmetrics.c (hunk 17 above)
Comment (mvollmer) [inline src/bridge/cockpitpcpmetrics.c]: Hmm, yes, agreed.  But we should show any error except ENOENT: When "/foo" exists, it should be a directory of archives.  If it doesn't exist, "/foo.meta" should exist instead.

**Comment by fche** [comment]
Hi -

> From pmGetConfig docs:

> If the variable is not found in either the environment or the PCP
> configuration file (or the PCP configuration file is not found and
> PCP_CONF is not set in the environment), then a fatal error message
> is printed and the process will exit. Although this sounds drastic,
> it is the only course of action available because the PCP
> configuration or installation is fatally flawed.

A flare for the dramatic, eh?  Please open a bug, this kind of
attitude/behavior is unacceptable from a -library-.


> Right, can't use that, can we?

In practice, you'll be OK; the pcp.conf file will be somewhere.


- FChE

**Comment by mvollmer** [comment]
> In practice, you'll be OK; the pcp.conf file will be somewhere.

But... PCP_DIR or PCP_CONF might need to be set but aren't for the bridge, etc.

**Comment by fche** [comment]
>     In practice, you'll be OK; the pcp.conf file will be somewhere.
>
> But... PCP_DIR or PCP_CONF might need to be set but aren't for the bridge, etc.

The pcp-conf subrpm provides /etc/pcp.conf and little else, and is required by
pcp-libs.  So in practice, you'll be OK.

**Comment by stefwalter** [comment]
Merged.

**Comment by mvollmer** [comment]
>>  Or else, remove the /var/lib/pcp/config/pmlogger/config.default file and restart the pmlogger service.
>
> That seems to be a good option.

The file contains this section at the bottom:

```
# DO NOT UPDATE THE FILE ABOVE THIS LINE
# Otherwise any changes may be lost the next time pmlogconf is
# used on this file.
#
# It is safe to make additions from here on ...
#

[access]
disallow .* : all;
disallow :* : all;
allow local:* : enquire;
```

That means we can't just remove it, right?

However, `pmlogger_check reload` seems to detect changes in `/var/lib/pcp/config/pmlogconf/...` and updates `config.default` as necessary.
